lib/logstash/inputs/lumberjack.rb

```diff
@@ -66,52 +69,50 @@ def run(output_queue)
     start_buffer_broker(output_queue)
 
     while true do
-      begin
-        # Wrapping the accept call into a CircuitBreaker
-        if @circuit_breaker.closed?
-          connection = @lumberjack.accept # Blocking call that creates a new connection
+      # Wrappingu the accept call into a CircuitBreaker
+      if @circuit_breaker.closed?
+        connection = @lumberjack.accept # Blocking call that creates a new connection
 
-          invoke(connection, codec.clone) do |_codec, line, fields|
-            _codec.decode(line) do |event|
+        invoke(connection, @codec.clone) do |_codec, line, fields|
+          _codec.decode(line) do |event|
+            begin
               decorate(event)
               fields.each { |k,v| event[k] = v; v.force_encoding(Encoding::UTF_8) }
-
-              @circuit_breaker.execute { @buffered_queue << event }
+              @circuit_breaker.execute { @buffered_queue.push(event, @timeout) }
+            rescue => e
+              raise e
             end
           end
-        else
-          @logger.warn("Lumberjack input: the pipeline is blocked, temporary refusing new connection.")
-          sleep(RECONNECT_BACKOFF_SLEEP)
         end
-        # When too many errors happen inside the circuit breaker it will throw 
-        # this exception and start refusing connection, we need to catch it but 
-        # it's safe to ignore.
-      rescue LogStash::CircuitBreaker::OpenBreaker => e
+      else
+        @logger.warn("Lumberjack input: the pipeline is blocked, temporary refusing new connection.")
+        sleep(RECONNECT_BACKOFF_SLEEP)
       end
     end
   rescue LogStash::ShutdownSignal
     @logger.info("Lumberjack input: received ShutdownSignal")
-  rescue => e
-    @logger.error("Lumberjack input: unhandled exception", :exception => e, :backtrace => e.backtrace)
   ensure
     shutdown(output_queue)
   end # def run
 
   private
-  def accept(&block)
-    connection = @lumberjack.accept # Blocking call that creates a new connection
-    block.call(connection, @codec.clone)
-  end
-
-  private
   def invoke(connection, codec, &block)
     @threadpool.post do
       begin
+        # If any errors occur in from the events the connection should be closed in the 
+        # library ensure block and the exception will be handled here
         connection.run do |fields|
           block.call(codec, fields.delete("line"), fields)
         end
-      rescue => e
-        @logger.error("Exception in lumberjack input thread", :exception => e)
+
+        # When too many errors happen inside the circuit breaker it will throw 
+        # this exception and start refusing connection, we need to catch it but 
+        # it's safe to ignore.
+      rescue LogStash::CircuitBreaker::OpenBreaker,
+        LogStash::CircuitBreaker::HalfOpenBreaker => e
+        logger.warn("Lumberjack input: connection closed, backing off", :exception => e.class)
```

Comment:
The way I read this message is taht "connection closed" is something that happened to this part of the machine, where as it seems more like that an OpenBreaker is causing us (here) to cause the closure directly. Thoughts on clarifying this? If the breaker is open, we want to tell users what caused it (slow pipeline?) and what we are doing about it (closing the connection intentionally)